libs/langgraph/src/agent/agent.ts

```diff
@@ -0,0 +1,203 @@
+import { LanguageModelLike } from "@langchain/core/language_models/base";
+import { type RunnableToolLike } from "@langchain/core/runnables";
+import type {
+  DynamicTool,
+  StructuredToolInterface,
+} from "@langchain/core/tools";
+import { ToolNode } from "../prebuilt";
+import { Annotation, StateGraph } from "../graph";
+import { END, MessagesAnnotation, START } from "../web";
+import {
+  BaseMessage,
+  isAIMessage,
+  SystemMessage,
+} from "@langchain/core/messages";
+import { unknown } from "zod/v4";
+import {
+  BaseChatModel,
+  ToolChoice,
+} from "@langchain/core/language_models/chat_models";
+import { EphemeralValue } from "../channels/ephemeral_value";
+
+type Tool = StructuredToolInterface | DynamicTool | RunnableToolLike;
+
+export const AgentState = Annotation.Root({
+  messages: MessagesAnnotation["spec"]["messages"],
+  jumpTo: new EphemeralValue<"tools" | "model" | "__end__">(),
+  response: Annotation<Record<string, unknown> | undefined>(),
+});
+
+export type AgentState = typeof AgentState;
+
+const isToolReturnDirect = (tool: Tool) =>
+  "returnDirect" in tool && tool.returnDirect;
+
+export interface AgentMiddleware {
```

Comment:
Doesn't this type require to have `state` and `context` generics so these are propagated to the hooks?

Reply:
also should we also include `beforeTool` and `afterTool`? Not that we specced this out before but seems to be a logical enhancements, maybe for later?